src/java.base/share/classes/java/security/PEMDecoder.java

```diff
@@ -0,0 +1,510 @@
+/*
+ * Copyright (c) 2025, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package java.security;
+
+import jdk.internal.javac.PreviewFeature;
+
+import sun.security.pkcs.PKCS8Key;
+import sun.security.rsa.RSAPrivateCrtKeyImpl;
+import sun.security.util.Pem;
+
+import javax.crypto.EncryptedPrivateKeyInfo;
+import javax.crypto.spec.PBEKeySpec;
+import java.io.*;
+import java.nio.charset.StandardCharsets;
+import java.security.cert.*;
+import java.security.spec.*;
+import java.util.Base64;
+import java.util.Objects;
+
+/**
+ * {@code PEMDecoder} implements a decoder for Privacy-Enhanced Mail (PEM) data.
+ * PEM is a textual encoding used to store and transfer security
+ * objects, such as asymmetric keys, certificates, and certificate revocation
+ * lists (CRLs).  It is defined in RFC 1421 and RFC 7468.  PEM consists of a
+ * Base64-formatted binary encoding enclosed by a type-identifying header
+ * and footer.
+ *
+ * <p> The {@linkplain #decode(String)} and {@linkplain #decode(InputStream)}
+ * methods return an instance of a class that matches the data
+ * type and implements {@link DEREncodable}.
+ *
+ * The following lists the supported PEM types and the {@code DEREncodable}
+ * types that each are decoded as:
+ * <ul>
+ *  <li>CERTIFICATE : {@code X509Certificate}</li>
+ *  <li>X509 CRL : {@code X509CRL}</li>
+ *  <li>PUBLIC KEY : {@code PublicKey}</li>
+ *  <li>PUBLIC KEY : {@code X509EncodedKeySpec} (Only supported when passed as a {@code Class} parameter)</li>
+ *  <li>PRIVATE KEY : {@code PrivateKey}</li>
+ *  <li>PRIVATE KEY : {@code PKCS8EncodedKeySpec} (Only supported when passed as a {@code Class} parameter)</li>
+ *  <li>PRIVATE KEY : {@code KeyPair} (if the encoding also contains a public key)</li>
```

Comment:
In a later paragraph you talk about reading `PublicKey` from a `PRIVATE KEY` if it is 2.0 and contains the public key. How about merging that info into this list?